src/shell/shell_batch.cpp

```diff
@@ -52,6 +52,19 @@ BatchFile::~BatchFile() {
 	shell->echo=echo;
 }
 
+// A helper function used in ReadLine. Copies the source to target provided the
+// result fits within CMD_MAXLINE, while taking the existing portion into
+// consideration. It moves the target ahead by the length copied.
+void append_cmd_write(char *&target, const char *existing, const char *source)
```

Comment:
Nice; (should be a static function).

Optional suggestion: how about making it a lambda expression inside the function? This way there's no need for documenting it as a helper usable in specific place only.

I would place it after line 116 or 117:
```C++
auto append_cmd_write = [&cmd_write, &line](const char *src) {
        // ...
}
```
If you want to avoid making it a lambda and stick to static function, that's fine by me as well :)

Reply:
Much nicer! Thanks for recommending it, and getting me started with the prototype!